### PR TITLE
chore(logger): pass log-mode to each controller and move function to

### PR DESCRIPTION
**Pull request by zdtsw**
a dedicated package

- move ConfigLoggers()  from common to new logger package
- add new function to convert log-mode string to int for V().
- update README with more detail

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: pkg/logger/logger.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+var logLevelMapping = map[string]int{
+	"devel":   0,
+	"default": 1, // default one when not set log-mode
+	"prod":    2,
+}
+
+// in each controller, to use different log level.
+func LogWithLevel(logger logr.Logger, level string) logr.Logger {
+	if len(level) == 0 {
```
Comment (ykaliuta) [inline pkg/logger/logger.go:19]: Probably, there is no sense for the check, the lookup check does the same.
Or may be make it non-mandatory argument (...string) and then check for the first member. But since it's an internal API, does not make a lot of sense.

File: pkg/logger/logger.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	verbosityLevel, ok := logLevelMapping[level]
+	if !ok {
+		verbosityLevel = 1 // fallback to info level
+	}
+	return logger.V(verbosityLevel)
+}
+
+// in DSC component, to use different mode for logging, e.g. development, production
+// when not set mode it falls to "default" which is used by startup main.go.
+func ConfigLoggers(mode string) logr.Logger {
```
Comment (ykaliuta) [inline pkg/logger/logger.go:31]: I'm wondering, with the planned log from context, will it change?

File: controllers/certconfigmapgenerator/certconfigmapgenerator_controller.go [hunk 3]
```diff
@@ -51,7 +51,7 @@ func (r *CertConfigmapGeneratorReconciler) Reconcile(ctx context.Context, req ct
 	// Get namespace instance
 	userNamespace := &corev1.Namespace{}
 	if err := r.Client.Get(ctx, client.ObjectKey{Name: req.Namespace}, userNamespace); err != nil {
-		return ctrl.Result{}, errors.WithMessage(err, "error getting namespace to inject trustedCA bundle ")
+		return ctrl.Result{}, errors.WithMessage(err, "error getting namespace to inject trustedCA bundle")
```
Comment (ykaliuta) [inline controllers/certconfigmapgenerator/certconfigmapgenerator_controller.go:54]: unrelated change (not a showstopper)

File: README.md [hunk 4]
```diff
@@ -228,13 +228,22 @@ This will ensure that the doc for the apis are updated accordingly.
...
+| prod             | error / 2           | highest level  |
+| production       | error / 2           | same as prod   |
+
 
 #### Component level
 
 Logger on components can be changed by DSCI devFlags during runtime.
 By default, if not set .spec.devFlags.logmode, it uses INFO level
-Modification applies to all components, not only these "Managed" ones
+Modification applies to all components, not only these "Managed" ones.
```
Comment (ykaliuta) [inline README.md:246]: unrelated change (not a showstopper)

File: README.md
Comment (ykaliuta) [inline README.md:1]: Hmm, wondering, how to comment on the commit message.
Looks like missing space after `move` and a typo in `looger`.

Comment (zdtsw) [reply README.md:1]: i will amend this in the message

File: pkg/logger/logger.go (hunk 2 above)
Comment (zdtsw) [reply pkg/logger/logger.go:31]: if you meant the proposal from last meeting, it might be a totally different way of how we handle loggers.
we might not need this function on the component level at all. but i am not sure what that solution will be.

Comment (zdtsw) [reply pkg/logger/logger.go:31]: i think here is only to move similar functionalities into one package, so we can better organize later

File: pkg/logger/logger.go (hunk 1 above)
Comment (zdtsw) [reply pkg/logger/logger.go]: you are right.
the original idea is to ensure, any " " will be chopped but i made a wrong code.
i did a bit change now to cover the cases 
- `"   devel" `is the same as `"devel"`, 
- `"          " `is the same as `""` as `"default"`

File: pkg/logger/logger.go (hunk 2 above)
Comment (ykaliuta) [reply pkg/logger/logger.go:30]: Yes, about that. Sure, fine to revisit later.

File: pkg/logger/logger.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+var logLevelMapping = map[string]int{
+	"devel":   0,
+	"default": 1, // default one when not set log-mode
+	"prod":    2,
+}
+
+// in each controller, to use different log level.
+func LogWithLevel(logger logr.Logger, level string) logr.Logger {
```
Comment (ykaliuta) [inline pkg/logger/logger.go:19]: will it check for "production" and "development" like doc says and ConfigLoggers does?

File: README.md [hunk 6]
```diff
@@ -249,7 +258,15 @@ spec:
 ```
 
 Avaiable value for logmode is "devel", "development", "prod", "production".
-The first two work the same set to DEBUG level; the later two work the same, using ERROR level.
+The first two work the same set to DEBUG level; the later two work the same as using ERROR level.
+
+| .spec.devFlags.logmode | stacktrace level | verbosity | Output   | Comments       |
+| ---------------------- | ---------------- | --------- | -------- | -------------- |
+| devel                  | WARN             | INFO      | Console  | lowest level, using epoch time  |
```
Comment (ykaliuta) [inline README.md:265]: Frankly, I'm confused here about verbosity level. So, by default it set to INFO ConfigLoggers. Then it can be changed by command line to operator or devflag in DSCI for components to debug, or how?

File: pkg/logger/logger.go (hunk 5 above)
Comment (zdtsw) [reply pkg/logger/logger.go:19]: i copy the line from "component" to the dsci level in README.
on one hand, i think good to have the same values on both places so user have a consistent experience.
on the other hand, it is a bit duplicated to have devel and development, prod and production here. 
Both are easy to change, but not sure which is better.

File: README.md (hunk 6 above)
Comment (zdtsw) [reply README.md:265]: i would say normally, user should not change anything,
and it uses info for both controllers and components.

but when something happens or user prefer more/less logging:
- if they want the change to be on controller: set the flag to operator, either in CSV or community user might wanna build their own image so they change code/config
- if they want the change to be on component: set the devflag.logmode in DSCI.  => i see this is more useful than the first one ^ esp. if it is running in production, a neat json for ETL, or a plaintext for debugging with stacktrace.

File: pkg/logger/logger.go (hunk 5 above)
Comment (ykaliuta) [reply pkg/logger/logger.go:19]: > i copy the line from "component" to the dsci level in README.

Yeah, but does it mean that "devel" sets the verbosity level, but "development" doesn't since lookup in the map fails?

Comment (zdtsw) [reply pkg/logger/logger.go:19]: yes, for the current code, it only support "" "devel" and "prod"
if we want to have "development" and "production", need to add these 2 into map.
or remove 2 lines in the README table.

File: README.md (hunk 6 above)
Comment (ykaliuta) [reply README.md:263]: > i would say normally, user should not change anything, and it uses info for both controllers and components.
> 
> but when something happens or user prefer more/less logging:
> 
> * if they want the change to be on controller: set the flag to operator, either in CSV or community user might wanna build their own image so they change code/config
> * if they want the change to be on component: set the devflag.logmode in DSCI.  => i see this is more useful than the first one ^ esp. if it is running in production, a neat json for ETL, or a plaintext for debugging with stacktrace.

But in the doc will the "verbosity" for "devel" here other than "INFO"?

Comment (zdtsw) [reply README.md:263]: not for what we already have in the code.
if INFO is not enough, (i tested DEBIG, it was loads out to stdout.) we can extend it in the future. 
but for now i will just keep INFO as the only verbositiy for devel, "" and "prod"

**Review by ykaliuta — APPROVED** [review]
